File: src/main/java/com/github/zxh/classpy/classfile/bytecode/NewArray.java
package com.github.zxh.classpy.classfile.bytecode;

import com.github.zxh.classpy.classfile.ClassParseException;
import com.github.zxh.classpy.classfile.reader.ClassReader;

public class NewArray extends Instruction {

    public NewArray(Opcode opcode, int pc) {
        super(opcode, pc);
    }
    
    @Override
    protected void readOperands(ClassReader reader) {
        int atype = reader.readU1().getValue();
        setDesc(getDesc() + " " + getArrayType(atype));
    }
    
    private static String getArrayType(int atype) {
        switch (atype) {
            case  4: return "boolean";
            case  5: return "char";
            case  6: return "float";
            case  7: return "double";
            case  8: return "byte";
            case  9: return "short";
            case 10: return "int";
            case 11: return "long";
            default: throw new ClassParseException("Invalid atype: " + atype);
        }
    }
    
}

File: src/main/java/com/github/zxh/classpy/classfile/bytecode/NewArray.kt
package com.github.zxh.classpy.classfile.bytecode

import com.github.zxh.classpy.classfile.ClassParseException
import com.github.zxh.classpy.classfile.constant.ConstantPool

class NewArray(opcode: Opcode, pc: Int) : Instruction(opcode, pc) {

    init {
        u1("opcode");
        u1("atype");
    }
    
    override fun afterRead(cp: ConstantPool) {
        val atype = getArrayType(super.getInt("atype"))
        desc = "${opcode.name} $atype"
    }
    
    private fun getArrayType(atype: Int): String {
        return when (atype) {
             4 -> "boolean";
             5 -> "char";
             6 -> "float";
             7 -> "double";
             8 -> "byte";
             9 -> "short";
            10 -> "int";
            11 -> "long";
            else -> throw ClassParseException("Invalid atype: " + atype);
        }
    }
    
}
